### PR TITLE
Add an in-memory state component

### PR DESCRIPTION
**Pull request by seeflood**
# Description

Add an in-memory state component. Similar to the `in-memory` pubsub component.

**This PR is co-authored by @skyao and me.** We did pair programming :)

## Issue reference
https://github.com/dapr/docs/issues/2014

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by CodeMonkeyLeet** [comment]
@seeflood thanks for contributing this, could you address the linter errors in the PR? 

Per the PR template, can you also open an issue in the dapr/docs repo to track that new documentation is added for this component? It should be very straightforward like the pubsub in-memory one.

**Comment by yaron2** [comment]
Thanks @seeflood this is great and can really help users use Dapr locally without Docker!

Please see @CodeMonkeyLeet's comment about the linter and docs.

**Comment by seeflood** [comment]
OK i will fix them.

>Thanks @seeflood this is great and can really help users use Dapr locally without Docker!
Yep,this is the purpose of this component, so that people who don’t have a docker environment can use dapr and even develop SDKs for dapr.

**Comment by mukundansundar** [comment]
@seeflood Thanks for this addition. As Yaron pointed out, this will help with local development in dapr. 
My questions are the following: 
1. How closely is this component going to support all the functionality that the other state stores give? For example given mismatching ETag, should the component delete the state or return error?
2. @yaron2 @artursouza Do you think this needs to support as many of the state store features as possible allowing for developers to use this in Dapr to develop locally and then switch another component? In that case shouldn't conformance tests be enabled for this component? 
3. My understanding is etag on insert/update(set) is generated by the  server and if needed can be used for conditional set, should that be supported here? 
The main question is how closely is the store going to conform to the Dapr state store API? 
//cc @yaron2 @artursouza

**Comment by yaron2** [comment]
> @seeflood Thanks for this addition. As Yaron pointed out, this will help with local development in dapr. My questions are the following:
> 
> 1. How closely is this component going to support all the functionality that the other state stores give? For example given mismatching ETag, should the component delete the state or return error?
> 2. @yaron2 @artursouza Do you think this needs to support as many of the state store features as possible allowing for developers to use this in Dapr to develop locally and then switch another component? In that case shouldn't conformance tests be enabled for this component?
> 3. My understanding is etag on insert/update(set) is generated by the  server and if needed can be used for conditional set, should that be supported here?
>    The main question is how closely is the store going to conform to the Dapr state store API?
>    //cc @yaron2 @artursouza

Ideally we have etag and transactions support for this store. As a dev tool, it needs to include all features of a state store.

**Comment by lotuc** [comment]
Would it be better if we create a sqlite store instead. This way the features MySQL & PostgreSQL supported would be easily supported and *it can be in-memory or locally persisted*. And with this the features proposed [here](https://github.com/dapr/dapr/issues/4289) may be implemented application level before it is officially supported by the dapr core.

**Comment by skyao** [comment]
@seeflood please update this PR and let's complete it.

@yaron2  @mukundansundar can we split this PR into two stage: 1. without advanced feature  2. with advanced feature.  Because in perf test for state, I need an in-memory state, thus I even submit an new PR (since I don't know this PR) https://github.com/dapr/components-contrib/issues/1657

If OK, we can quickly merge this PR (or my PR) and don't block perf test of state.

**Comment by seeflood** [comment]
sorry for the delay. I will update this PR this week.
I agree with @skyao ,can we split this component into two pr?

**Comment by mukundansundar** [comment]
> @seeflood please update this PR and let's complete it.
> 
> @yaron2 @mukundansundar can we split this PR into two stage: 1. without advanced feature 2. with advanced feature. Because in perf test for state, I need an in-memory state, thus I even submit an new PR (since I don't know this PR) #1657
> 
> If OK, we can quickly merge this PR (or my PR) and don't block perf test of state.

@skyao @seeflood I am ok with splitting it into two PRs, will leave it to the maintainers to take the final call. 
@berndverst @yaron2 @artursouza any concerns with that?

**Comment by daixiang0** [comment]
I am ok with splitting it into two PRs.

File: state/in-memory/in_memory.go
Comment (daixiang0) [inline state/in-memory/in_memory.go:6]: Please fix lint issues.

File: state/in-memory/in_memory.go
Comment (daixiang0) [inline state/in-memory/in_memory.go:22]: How about using` sync.map`?

**Comment by skyao** [comment]
@yaron2 @daixiang0 @berndverst @mukundansundar please help to review this PR. I have worked with @seeflood, now the implementation of in memory state has most of the state feature including etag, transactional, expire. Yet state query is not not included since there is no requirement for this feature.

Now perf test of state is blocked by this PR.

File: state/in-memory/in_memory.go
Comment (daixiang0) [reply state/in-memory/in_memory.go:36]: A lock controlled by us is better in this case, let's keep it.

File: state/in-memory/in_memory_test.go
Comment (daixiang0) [inline state/in-memory/in_memory_test.go:36]: Please rename it as happy or bad cases and cover them.

File: state/in-memory/in_memory_test.go
Comment (daixiang0) [inline state/in-memory/in_memory_test.go:46]: Same as above.

**Comment by daixiang0** [comment]
> @yaron2 @daixiang0 @berndverst @mukundansundar please help to review this PR. I have worked with @seeflood, now the implementation of in memory state has most of the state feature including etag, transactional, expire. Yet state query is not not included since there is no requirement for this feature.
> 
Please create an issue to track it.

**Comment by daixiang0** [comment]
@berndverst I often see certification tests like rabbitmq or kafka flaky, have we tracked it?

File: state/in-memory/in_memory.go
Comment (mukundansundar) [inline state/in-memory/in_memory.go:107]: Can this be replaced by `DefaultBulkStateStore` Implementation of `BulkDelete`?

File: state/in-memory/in_memory.go
Comment (mukundansundar) [inline state/in-memory/in_memory.go:176]: Can this be replaced by `DefaultBulkStateStore` Implementation of `BulkGet`?

File: state/in-memory/in_memory.go
Comment (mukundansundar) [inline state/in-memory/in_memory.go:250]: Can this be replaced by `DefaultBulkStateStore` Implementation of `BulkSet`?

File: state/in-memory/in_memory.go
Comment (mukundansundar) [inline state/in-memory/in_memory.go:372]: Can this be added in this PR?

**Review by mukundansundar — COMMENTED** [review]
Can the Bulk operations use the `store.DetaultBulkStore` implementation of them? I am not seeing that much difference in what has been implemented from the default implementation. 
Also can conformance tests be added for this store?

File: state/in-memory/in_memory.go
Comment (Taction) [inline state/in-memory/in_memory.go:95]: Go pointers values are only equal when they point to the same value in the memory or if they are nil. 
```
if  *item.etag != *etag {
```
If item.etag is possibly nil, it should be checked.

File: state/in-memory/in_memory.go
Comment (Taction) [inline state/in-memory/in_memory.go:167]: This func should return the item and a bool value to indicate wether it is deleted(or just the item, null for deleted). If the item is not deleted, `Get` should returns the item instead of nil.

**Comment by skyao** [comment]
> 

Why not use defaultBulkStore is for performance. 

We don't want to get lock / unlock for each items in the bulk list. 

So these three bulk method are follow the same strategy: 1. first check input parameter without lock, if fail, return error, no lock in this fail case. 2. get lock and do etag check for all items  3. do save or delete for all items 4. unlock

step2 and step 3 are protected by a BIG lock for all items.

File: state/in-memory/in_memory_test.go
Comment (seeflood) [reply state/in-memory/in_memory_test.go]: fixed

File: state/in-memory/in_memory_test.go
Comment (seeflood) [reply state/in-memory/in_memory_test.go]: fixed

File: state/in-memory/in_memory.go
Comment (seeflood) [reply state/in-memory/in_memory.go]: fixed

File: state/in-memory/in_memory.go
Comment (seeflood) [reply state/in-memory/in_memory.go]: fixed

File: state/in-memory/in_memory.go
Comment (seeflood) [reply state/in-memory/in_memory.go]: done

**Comment by seeflood** [comment]
All resolved. Please review again 🤔

**Comment by seeflood** [comment]
> > @yaron2 @daixiang0 @berndverst @mukundansundar please help to review this PR. I have worked with @seeflood, now the implementation of in memory state has most of the state feature including etag, transactional, expire. Yet state query is not not included since there is no requirement for this feature.
> 
> Please create an issue to track it.

@daixiang0 Ok.  I created https://github.com/dapr/components-contrib/issues/1705

**Comment by skyao** [comment]
@Taction @daixiang0 @mukundansundar  Please help to review and merge this PR. Thank you. My PR of state perf test has been blocked by this PR for more than one month......

**Comment by skyao** [comment]
@seeflood Please update the branch.

File: state/in-memory/in_memory_test.go
Comment (Taction) [inline state/in-memory/in_memory_test.go:139]: Duplicated with L111.

File: state/in-memory/in_memory.go
Comment (Taction) [inline state/in-memory/in_memory.go:114]: `if item.etag == nil` this format causes panic due to the usage of `*item.etag`.

**Review by Taction — DISMISSED** [review]
Thanks for your contribution! LGTM overall. Just some nits changes.

File: state/in-memory/in_memory_test.go
Comment (seeflood) [reply state/in-memory/in_memory_test.go]: ok. deleted

File: state/in-memory/in_memory.go
Comment (seeflood) [reply state/in-memory/in_memory.go:114]: we already checked nil at L109 :
image

So it's safe here.

Comment (Taction) [reply state/in-memory/in_memory.go:114]: What I would say is that `item.etag` is probably nil, not item. And `fmt.Errorf` uses `*item.etag` without check.

Comment (seeflood) [reply state/in-memory/in_memory.go]: @Taction Thanks. 
fixed

**Comment by seeflood** [comment]
All review issues resolved.
@yaron2 @artursouza Hi, could u help merge this PR?
@skyao needs this component :)

**Comment by seeflood** [comment]
All review conversations resolved.
@yaron2 @artursouza @berndverst  Hi, could u help merge this PR?
@skyao needs this component to do perf test :)

**Review by Taction — APPROVED** [review]
LGTM

**Comment by yaron2** [comment]
Excellent work @seeflood and @skyao. Kindly open a docs PR: https://github.com/dapr/docs/issues/2014.

**Comment by berndverst** [comment]
FYI: This component never passed conformance tests. Per my request here https://github.com/dapr/components-contrib/pull/1326#issuecomment-1067150741 conformance tests should have been added but unfortunately my comments were ignored as I was not a maintainer yet.

@ItalyPaleAle is now looking into fixing this component in a separate PR.

**Comment by yaron2** [comment]
> FYI: This component never passed conformance tests. Per my request here [#1326 (comment)](https://github.com/dapr/components-contrib/pull/1326#issuecomment-1067150741) conformance tests should have been added but unfortunately my comments were ignored as I was not a maintainer yet.
> 
> @ItalyPaleAle is now looking into fixing this component in a separate PR.

Great that we are adding conformance tests for alpha components.

I'll just say that your comment requests regarding the work needed to merge the PR were carried out to the fullest by the author, as you asked for etag and transaction support and both have been added, while the conformance tests were a recommendation to the author.

**Comment by berndverst** [comment]
Yes, my comment is directed towards maintainers more so. When possible let's ensure conformance tests are enabled.

No big deal here, though surprisingly we didn't catch this component wasn't conformant until now.

**Comment by yaron2** [comment]
> Yes, my comment is directed towards maintainers more so. When possible let's ensure conformance tests are enabled.
> 
> No big deal here, though surprisingly we didn't catch this component wasn't conformant until now.

I can't speak on behalf of other maintainers, but personally I'd have asked to dismiss the approval had a conformance test been required instead of recommended.

Basically the case where a conformance test can't be added should be rare, so maybe we can mandate it as part of the acceptance criteria and update the PR template?

**Comment by ItalyPaleAle** [comment]
I think that for certain kinds of components, we should require that a conf test be set up. When we don’t have the infrastructure to run it in our CI (e.g. requires an account with a  cloud provider we don’t support), we should ask that the contributor at least runs the conformance test manually before submitting the component and validate they pass.

In this specific case, for example, [attempting to set up the conf test ](https://github.com/dapr/components-contrib/pull/2303) revealed a very large number of errors, from runtime panics due to incorrect type assertions, to bad handling of JSON-encoded data, to incorrect ETag and TTL implementations. Something that should have arguably been caught much earlier.

**Pull request merged**
